File: leetcode/0256-paint-house/Solution.java
class Solution {
    public int minCost(int[][] costs) {
        int red = costs[0][0], blue = costs[0][1], green = costs[0][2];

        for (int i = 1; i < costs.length; i++) {
            int newRed = costs[i][0] + Math.min(blue, green);
            int newBlue = costs[i][1] + Math.min(red, green);
            int newGreen = costs[i][2] + Math.min(red, blue);
            red = newRed;
            blue = newBlue;
            green = newGreen;
        }

        return Math.min(red, Math.min(blue, green));
    }
}

File: leetcode/0256-paint-house/solution.py
from typing import List


class Solution:
    def minCost(self, costs: List[List[int]]) -> int:
        red, blue, green = costs[0]

        for i in range(1, len(costs)):
            new_red = costs[i][0] + min(blue, green)
            new_blue = costs[i][1] + min(red, green)
            new_green = costs[i][2] + min(red, blue)
            red, blue, green = new_red, new_blue, new_green

        return min(red, blue, green)
